uwnet/stochastic_parameterization/utils.py

```diff
@@ -0,0 +1,200 @@
+import numpy as np
+from functools import lru_cache
+import torch
+from uwnet.tensordict import TensorDict
+import xarray as xr
+
+model_dir = ''
+# model_dir = '/Users/stewart/projects/uwnet/uwnet/stochastic_parameterization/'  # noqa
+model_location = model_dir + 'stochastic_model.pkl'
+default_base_model_location = model_dir + 'full_model/1.pkl'
+default_ds_location = "training.nc"
+# default_ds_location = "uwnet/stochastic_parameterization/training.nc"
+dataset_dt_seconds = 10800
+# default_binning_method = 'precip'
+default_binning_method = 'column_integrated_qt_residuals'
+default_binning_quantiles = (0.06, 0.15, 0.30, 0.70, 0.85, 0.94, 1)
+# default_binning_quantiles = (.1, .3, .7, .9, 1)
+# default_binning_quantiles = (1,)
+# default_binning_quantiles = (.01, .05, .15, .35, .5, .65, .85, .95, .99, 1)
+
+
+class BaseModel(object):
+
+    def __init__(
+            self,
+            model_location,
+            dataset,
+            binning_quantiles=default_binning_quantiles,
+            time_step_days=.125,
+            binning_method=default_binning_method):
+        self.model = torch.load(model_location)
+        self.ds = dataset
+        self.time_step_days = time_step_days
+        self.time_step_seconds = 86400 * time_step_days
+        self.binning_quantiles = binning_quantiles
+        self.binning_method = binning_method
+
+    def get_true_nn_forcing_idx(self, time_idx):
+        start = self.ds.isel(time=time_idx)
+        stop = self.ds.isel(time=time_idx + 1)
+        true_nn_forcing = {}
+        for key in ['QT', 'SLI']:
+            forcing = (start[f'F{key}'].values + stop[f'F{key}'].values) / 2
```

Comment:
Most of the methods in this class fit a pattern: they take xarray inputs, immediately call `.values`, do some calculations, and then return numpy outputs. It would be cleaner if all these actions were performed with `xarray`.  If that is not possible, then they should all take numpy inputs,  and the xarray--numpy conversion be isolated to `insert_nn_output_precip_ratio_bin_membership`. In general, I would avoid `.values`.